common/src/worldgen.rs

```diff
@@ -167,154 +169,132 @@ impl ChunkParams {
+
+                    // elev is calculated in multiple steps. The initial value elev_pre_terracing
+                    // is used to calculate elev_pre_noise which is used to calculate elev.
+                    let elev_pre_terracing = trilerp(&self.env.max_elevations, cube_coords);
+                    let block = trilerp(&self.env.blockinesses, cube_coords);
+                    let voxel_elevation = self.surface.distance_to_chunk(self.chunk, &center);
+                    let strength = 0.4 / (1.0 + voxel_elevation.powi(2));
+                    let terracing_small =
+                        terracing_diff(elev_pre_terracing, block, 5.0, strength, 2.0);
+                    let terracing_big =
+                        terracing_diff(elev_pre_terracing, block, 15.0, strength, -1.0);
+                    // Small and big terracing effects must not sum to more than 1,
+                    // otherwise the terracing fails to be (nonstrictly) monotonic
+                    // and the terrain gets trenches ringing around its cliffs.
+                    let elev_pre_noise =
+                        elev_pre_terracing + 0.6 * terracing_small + 0.4 * terracing_big;
+
+                    // initial value dist_pre_noise is the difference between the voxel's distance
+                    // from the guiding plane and the voxel's calculated elev value. It represents
+                    // how far from the terrain surface a voxel is.
+                    let dist_pre_noise = elev_pre_noise / TERRAIN_SMOOTHNESS - voxel_elevation;
+
+                    // adding noise allows interfaces between strata to be rough
+                    let elev = elev_pre_noise + TERRAIN_SMOOTHNESS * rng.sample(&normal.unwrap());
+
+                    // Final value of dist is calculated in this roundabout way for greater control
+                    // over how noise in elev affects dist.
+                    let dist = if dist_pre_noise > 0.0 {
+                        // The .max(0.0) keeps the top of the ground smooth
+                        // while still allowing the surface/general terrain interface to be rough
+                        (elev / TERRAIN_SMOOTHNESS - voxel_elevation).max(0.0)
+                    } else {
+                        // Distance not updated for updated elevation if distance was originally
+                        // negative. This ensures that no voxels that would have otherwise
+                        // been void are changed to a material---so no floating dirt blocks.
+                        dist_pre_noise
+                    };
+
+                    let voxel_mat = if dist >= 0.0 {
+                        VoronoiInfo::terraingen_voronoi(elev, rain, temp, dist)
+                    } else {
+                        Material::Void
+                    };
+
+                    voxels.data_mut(self.dimension)[index(self.dimension, coords)] = voxel_mat;
```

Comment:
`data_mut` implicitly allocates memory for the voxel chunk, so we shouldn't call it unless we know we're going to assign something other than `Void`, to avoid wasting space on the empty chunks that the short circuit didn't catch.

Reply:
I've now made it so data_mut() is only called when voxel_mat has been assigned a non-void material.